### PR TITLE
Added endpoint to proxy the kube api

### PR DESCRIPTION
**Pull request by MrSupiri**
There is a custom web server running in port 9090, with this PR I have created an endpoint on `http://operator:9090/k8s/` which will forward all the traffic coming to it to a  kubectl proxy at port 8231 after dropping `/k8s` from the request URL and which itself will forward the request to kube-api after inserting the bearer token taken from the operators' service account.

### CODE REVIEW

File: operator/config/manager/manager.yaml [hunk 1]
```diff
@@ -25,32 +25,35 @@ spec:
...
-          periodSeconds: 10
-        resources:
-          limits:
-            cpu: 100m
-            memory: 30Mi
-          requests:
-            cpu: 100m
-            memory: 20Mi
+        - args: ["proxy", "--port=8231"]
+          image: bitnami/kubectl:latest
```
Comment (MrSupiri) [inline operator/config/manager/manager.yaml:29]: Here I am using [bitnami's image](https://hub.docker.com/r/bitnami/kubectl/) which has more than 1 billion pulls and the build files are open sources https://github.com/bitnami/bitnami-docker-kubectl and they seem to using a bot to automatically update the images when there is a new version of kubectl. Is this ok or should I build another custom image?

File: operator/Dockerfile [hunk 2]
```diff
@@ -13,13 +13,15 @@ RUN go mod download
...
 COPY controllers/ controllers/
+COPY internal/ internal/
 
 # Build
 RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -a -o manager main.go
 
 # Use distroless as minimal base image to package the manager binary
 # Refer to https://github.com/GoogleContainerTools/distroless for more details
-FROM gcr.io/distroless/static:nonroot
+#FROM gcr.io/distroless/static:nonroot
```
Comment (paynejacob) [inline operator/Dockerfile:23]: we can drop alpine, for production distroless makes sense

File: operator/Makefile [hunk 3]
```diff
@@ -67,7 +67,8 @@ docker-build: test ## Build docker image with the manager.
 	docker build -t ${IMG} .
 
 docker-push: ## Push docker image with the manager.
-	docker push ${IMG}
+	docker tag ${IMG} localhost:37129/${IMG}
```
Comment (paynejacob) [inline operator/Makefile:70]: this is very specific to your env but we are going to be redoing all this next sprint so we can leave it for now.

File: operator/config/manager/manager.yaml (hunk 1 above)
Comment (paynejacob) [reply operator/config/manager/manager.yaml:29]: I would rather we download kubectl into our image, that way airgap users only have to download 1 image.  We can leave this for the work next sprint.

File: operator/internal/server/handlers.go [hunk 4]
```diff
@@ -18,6 +22,17 @@ func (ws *WebServer) WriteResponse(w http.ResponseWriter, httpRes HTTPResponse)
 	}
 }
 
+func (ws *WebServer) ProxyToKubeAPI(res http.ResponseWriter, req *http.Request) {
+	targetUrl, _ := url.Parse("http://localhost:8231")
+	// create the reverse proxy
+	proxy := httputil.NewSingleHostReverseProxy(targetUrl)
+
+	req.URL.Path = "/" + strings.Join(strings.Split(req.URL.Path, "/")[2:], "/")
```
Comment (paynejacob) [inline operator/internal/server/handlers.go:30]: can you test <host>/k8s and see if you get an index error?

**Review by paynejacob — CHANGES_REQUESTED** [review]

File: operator/internal/server/handlers.go (hunk 4 above)
Comment (MrSupiri) [reply operator/internal/server/handlers.go:30]: image
Seems to work fine

File: operator/Makefile (hunk 3 above)
Comment (MrSupiri) [reply operator/Makefile:70]: ok I will open issue for this

**Comment by MrSupiri** [comment]
As our issue was with URL routing I revert back to the old technique which will just pull the token from the mount and send the request directly to kube-api. Which will remove the need for the kubectl proxy.

image

File: operator/internal/server/server.go [hunk 5]
```diff
@@ -42,6 +90,7 @@ func (ws *WebServer) ListenAndServe(stopCh <-chan struct{}) {
 
 	r.HandleFunc("/log-output/deploy", ws.DeployLogOutput).Methods("POST")
```
Comment (paynejacob) [inline operator/internal/server/server.go:91]: can we remove these if they are unused?

Comment (MrSupiri) [reply operator/internal/server/server.go]: yes we can

File: operator/internal/server/handlers.go [hunk 6]
```diff
@@ -1,85 +1,32 @@
...
-	}
+	proxy := httputil.NewSingleHostReverseProxy(targetUrl)
+	proxy.Transport = ws.kubeProxy.proxyTransport
+	req.URL.Path = "/" + strings.Join(strings.Split(req.URL.Path, "/")[2:], "/")
+	req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", ws.kubeProxy.token))
 
-	w.WriteHeader(200)
-	ws.WriteResponse(w, HTTPResponse{Data: &HTTPData{Message: "log-output destroyed successfully"}})
-}
+	ws.logger.V(1).Info(fmt.Sprintf("proxying %s to %s", req.URL, targetUrl))
```
Comment (paynejacob) [inline operator/internal/server/handlers.go:30]: can we lower this to debug? request logging can get noisy

**Review by paynejacob — APPROVED** [review]

**Pull request merged**